src/libstore/aws-creds.cc

```diff
@@ -0,0 +1,246 @@
+        }
+    }
+
+    AwsCredentialProvider(std::shared_ptr<Aws::Crt::Auth::ICredentialsProvider> provider)
+        : provider(provider)
+    {
+    }
+
+    AwsCredentials getCredentials()
+    {
+        if (!provider || !provider->IsValid()) {
+            throw AwsAuthError("AWS credential provider is invalid");
+        }
+
+        struct State
+        {
+            std::optional<AwsCredentials> result;
+            int resolvedErrorCode = 0;
+            bool resolved = false;
+        };
+
+        Sync<State> state;
+        std::condition_variable cv;
+
+        provider->GetCredentials([&](std::shared_ptr<Aws::Crt::Auth::Credentials> credentials, int errorCode) {
+            auto state_ = state.lock();
+
+            if (errorCode != 0 || !credentials) {
+                state_->resolvedErrorCode = errorCode;
+            } else {
+                auto accessKeyId = credentials->GetAccessKeyId();
+                auto secretAccessKey = credentials->GetSecretAccessKey();
+                auto sessionToken = credentials->GetSessionToken();
+
+                std::optional<std::string> sessionTokenStr;
+                if (sessionToken.len > 0) {
+                    sessionTokenStr = std::string(reinterpret_cast<const char *>(sessionToken.ptr), sessionToken.len);
+                }
+
+                state_->result = AwsCredentials(
+                    std::string(reinterpret_cast<const char *>(accessKeyId.ptr), accessKeyId.len),
+                    std::string(reinterpret_cast<const char *>(secretAccessKey.ptr), secretAccessKey.len),
+                    sessionTokenStr);
+            }
+
+            state_->resolved = true;
+            cv.notify_one();
+        });
+
+        {
+            auto state_ = state.lock();
+            // AWS CRT GetCredentials is asynchronous and only guarantees the callback will be
+            // invoked if the initial call returns success. There's no documented timeout mechanism,
+            // so we add a timeout to prevent indefinite hanging if the callback is never called.
+            auto timeout = std::chrono::seconds(30);
+            if (!state_.wait_for(cv, timeout, [&] { return state_->resolved; })) {
+                throw AwsAuthError(
+                    "Timeout waiting for AWS credentials (%d seconds)",
+                    std::chrono::duration_cast<std::chrono::seconds>(timeout).count());
+            }
+        }
+
+        auto state_ = state.lock();
+        if (!state_->result) {
+            throw AwsAuthError("Failed to resolve AWS credentials: error code %d", state_->resolvedErrorCode);
+        }
+
+        return *state_->result;
+    }
+};
+
+// Credential provider cache
+struct CredentialProviderCache
+{
+    // Key: profile name (empty string for default profile)
+    std::map<std::string, std::shared_ptr<AwsCredentialProvider>> providers;
+};
+
+// Global credential provider cache
+Sync<CredentialProviderCache> credentialProviderCache;
```

Comment:
This is hardly performance sensitive, but we recently switched to using Boost's concurrent hash map for this sort of thing, and we could use it for this too.